addons/mail/controllers/discuss.py

```diff
@@ -87,6 +88,8 @@ def _response_discuss_channel_invitation(self, channel_sudo, is_channel_token_se
                     channel_sudo = channel_sudo.with_context(guest=guest)
                     channel_sudo.add_members(guest_ids=[guest.id])
                 else:
+                    if channel_sudo.public == 'groups':
+                        raise UserError(_('Not Found'))
```

Comment:
Indeed, this change should be done.

Note that the reason for using 404 instead of user error in the controller is to avoid leaking information, as this is a public route. A clear error would indicate to potential attackers that this channel does exist.

Note that as we want to "hide" the real error from public routes, there should be a try/except on the `UserError` from inside the controllers when calling add members, which then raises a 404.